big/2/public-tests/tests/mikolajkowe.rs

```diff
@@ -0,0 +1,264 @@
+
+    // then
+    const EXPECTED_RESPONSES_SIZE: usize = 48;
+    let mut buf = [0_u8; EXPECTED_RESPONSES_SIZE];
+    stream
+        .read_exact(&mut buf)
+        .await
+        .expect("Less data then expected");
+
+    // asserts for write response
+    assert_eq!(&buf[0..4], MAGIC_NUMBER.as_ref());
+    assert_eq!(buf[6], 0x1); /* Invalid hmac */
+    assert_eq!(buf[7], 0x42);
+    assert_eq!(
+        u64::from_be_bytes(buf[8..16].try_into().unwrap()),
+        request_identifier
+    );
+    assert!(hmac_tag_is_ok(&hmac_client_key, &buf));
+}
+
+#[tokio::test]
+#[timeout(4000)]
+async fn bad_sector_number() {
+    // given
+    log_init();
+    let hmac_client_key = [5; 32];
+    let tcp_port = 30_287;
+    let storage_dir = tempdir().unwrap();
+    let request_identifier = 1778;
+
+    let config = Configuration {
+        public: PublicConfiguration {
+            tcp_locations: vec![("127.0.0.1".to_string(), tcp_port)],
+            self_rank: 1,
+            n_sectors: 20,
+            storage_dir: storage_dir.into_path(),
+        },
+        hmac_system_key: [1; 64],
+        hmac_client_key,
+    };
+
+    tokio::spawn(run_register_process(config));
+
+    tokio::time::sleep(Duration::from_millis(300)).await;
+    let mut stream = TcpStream::connect(("127.0.0.1", tcp_port))
+        .await
+        .expect("Could not connect to TCP port");
+    let write_cmd = RegisterCommand::Client(ClientRegisterCommand {
+        header: ClientCommandHeader {
+            request_identifier,
+            sector_idx: 20,
+        },
+        content: ClientRegisterCommandContent::Write {
+            data: SectorVec(vec![3; 4096]),
+        },
+    });
+
+    // when
+    send_cmd(&write_cmd, &mut stream, &hmac_client_key).await;
+
+    // then
+    const EXPECTED_RESPONSES_SIZE: usize = 48;
+    let mut buf = [0_u8; EXPECTED_RESPONSES_SIZE];
+    stream
+        .read_exact(&mut buf)
+        .await
+        .expect("Less data then expected");
+
+    // asserts for write response
+    assert_eq!(&buf[0..4], MAGIC_NUMBER.as_ref());
+    assert_eq!(buf[6], 0x2);
+    assert_eq!(buf[7], 0x42);
+    assert_eq!(
+        u64::from_be_bytes(buf[8..16].try_into().unwrap()),
+        request_identifier
+    );
+    assert!(hmac_tag_is_ok(&hmac_client_key, &buf));
+}
+
+#[tokio::test]
+#[serial_test::serial]
+#[timeout(30000)]
+async fn concurrent_writes_are_serialized() {
+    // given
+    log_init();
+    let port_range_start = 21518;
+    let n_clients = 16;
+    /* Spawn two and add our stub to system */
+    let config = TestProcessesConfig::new(3, port_range_start);
+    tokio::spawn(run_register_process(config.config(0)));
+    tokio::spawn(run_register_process(config.config(1)));
+    let listener = TcpListener::bind(config.tcp_locations[2].clone())
+        .await
+        .unwrap();
+    tokio::time::sleep(Duration::from_millis(300)).await;
+
+    let (mut receiver, _addr) = listener.accept().await.unwrap();
```

Comment:
I disagree with the assumption that register processes have to create connections between each other at startup.

I suggest moving it below sending all Write requests.
It would allow processes to connect to each other during client request handling.

Reply:
I will merge when this is fixed.

Reply:
You have to accept incoming connections always. What if that was a client?

Reply:
Oh, my bad xd
That's a listener

Reply:
Fixed